File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/DefaultHttpClient.java
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.

package com.microsoft.aspnet.signalr;

import java.io.IOException;
import java.util.concurrent.CompletableFuture;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;

class DefaultHttpClient extends HttpClient {
    private OkHttpClient client = new OkHttpClient();

    @Override
    public CompletableFuture<HttpResponse> send(HttpRequest httpRequest) {
        Request.Builder requestBuilder = new Request.Builder().url(httpRequest.url);
        if (httpRequest.method == "GET") {
            requestBuilder.get();
        } else if (httpRequest.method == "POST") {
            RequestBody body = RequestBody.create(null, new byte[] {});
            requestBuilder.post(body);
        } else if (httpRequest.method == "DELETE") {
            requestBuilder.delete();
        }

        Request request = requestBuilder.build();

        CompletableFuture<HttpResponse> responseFuture = new CompletableFuture<>();

        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                responseFuture.completeExceptionally(e.getCause());
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                try (ResponseBody body = response.body()) {
                    HttpResponse httpResponse = new HttpResponse(response.code(), response.message(), body.string());
                    responseFuture.complete(httpResponse);
                }
            }

        });

        return responseFuture;
    }
}

File: clients/ts/signalr/src/DefaultHttpClient.ts
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.

import { AbortError } from "./Errors";
import { HttpClient, HttpRequest, HttpResponse } from "./HttpClient";
import { ILogger } from "./ILogger";
import { XhrHttpClient } from "./XhrHttpClient";

let nodeHttpClientModule: any;
if (typeof XMLHttpRequest === "undefined") {
    // tslint:disable-next-line:no-var-requires
    nodeHttpClientModule = require("./NodeHttpClient");
}

/** Default implementation of {@link @aspnet/signalr.HttpClient}. */
export class DefaultHttpClient extends HttpClient {
    private readonly httpClient: HttpClient;

    /** Creates a new instance of the {@link @aspnet/signalr.DefaultHttpClient}, using the provided {@link @aspnet/signalr.ILogger} to log messages. */
    public constructor(logger: ILogger) {
        super();

        if (typeof XMLHttpRequest !== "undefined") {
            this.httpClient = new XhrHttpClient(logger);
        } else if (typeof nodeHttpClientModule !== "undefined") {
            this.httpClient = new nodeHttpClientModule.NodeHttpClient(logger);
        } else {
            throw new Error("No HttpClient could be created.");
        }
    }

    /** @inheritDoc */
    public send(request: HttpRequest): Promise<HttpResponse> {
        // Check that abort was not signaled before calling send
        if (request.abortSignal && request.abortSignal.aborted) {
            return Promise.reject(new AbortError());
        }

        if (!request.method) {
            return Promise.reject(new Error("No method defined."));
        }
        if (!request.url) {
            return Promise.reject(new Error("No url defined."));
        }

        return this.httpClient.send(request);
    }

    // @ts-ignore
    private getCookies(url: string): string {
        if ((this.httpClient as any).getCookies) {
            return (this.httpClient as any).getCookies(url);
        }
        return "";
    }
}
